### PR TITLE
[consensus] Add ReputationScoreCalculator & ScoringStrategy

### PR DESCRIPTION
**Pull request by arun-koshy**
## Description 
This is one of many PRs to implement Leader Scoring & Leader Schedule feature for Mysticeti.
- [Design Doc](https://www.notion.so/mystenlabs/Leader-Schedule-Scoring-538c65cb370b4644944a94485efc4979?pvs=4)
- [Prototype PR](https://github.com/MystenLabs/sui/pull/16975)

Notes:
- Unscored subdags will eventually be cached in `DagState` and when there are enough commits to trigger a leader schedule change the `ReputationScoreCalculator` will be used to calculate the `ReputationScores`. As can be seen in the flow diagram below.
- Added multiple scoring strategies so that we can experiment with them quickly until we decided on the strategy that is best to use. By default we will use the `VoteScoringStrategy` when everything is hooked up as it had the best early [results](https://metrics.sui.io/goto/yeUlNPxIR?orgId=1).
- We are sharing the `UniversalCommitter` inside of the `ReputationScoreCalculator` so we have consistency with the leaders that we elect when scoring.

Screenshot 2024-04-03 at 7 59 09 PM

## Test Plan 

unit tests

### CODE REVIEW

File: consensus/core/src/leader_scoring.rs
Comment (arun-koshy) [reply consensus/core/src/leader_scoring.rs:216]: We are just trying to get the rounds in which there may potentially be a leader and we do this by getting all the rounds of blocks that exist in the `unscored_subdag`. Then when we go to calculate scores for a leader because we are looping through the committer we will get the expected leaders within the same round because the leader offset is a part of the committer.

I will use a set so we filter out the duplicate rounds though

File: consensus/core/src/leader_scoring_strategy.rs
Comment (arun-koshy) [reply consensus/core/src/leader_scoring_strategy.rs:109]: Yes that is exactly right, we are just trying to add an additional layer on top of the vote scoring to indicate the quality of the vote. The thinking is that votes can easily get generated but actually distributing those votes is what is useful for the network. Additionally we see the best score [distribution](https://metrics.sui.io/goto/L5oyXFBSR?orgId=1) with `CertifiedVoteScoring`

File: consensus/core/src/leader_scoring_strategy.rs
Comment (arun-koshy) [reply consensus/core/src/leader_scoring_strategy.rs:291]: I agree I think we need to conclude on one, but the reason I kept them all was so that we could get the feature in and then test them all thoroughly later and choose a strategy with more information. Do you want me to remove it now or are you okay merging with them all for now.

File: consensus/core/src/leader_scoring_strategy.rs
Comment (arun-koshy) [reply consensus/core/src/leader_scoring_strategy.rs:372]: Yeah I was waiting for the DagBuilder code to land before I write those tests. After I merge all the scoring/schedule work I will make sure to add those more complex scenarios in tests

File: consensus/core/src/leader_scoring.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring.rs:195]: Maybe `UnscoredBlocks` since they are what are contained.

**Comment by mwtian** [comment]
Still reviewing, I will leave comments incrementally.

File: consensus/core/src/leader_scoring.rs
Comment (arun-koshy) [reply consensus/core/src/leader_scoring.rs]: Changed it so that when an UnscoredSubdag is created the list of subdags that are passed in will be checked to ensure that we have a contiguous (and therefore sorted) list of commit indices. Then we will also have the blocks collected in order as well so we can get the commit range and leader range easily from the UnscoredSubdag.

> The result from minmax() is inclusive for both start and end. So CommitRange needs to end on *max_commit_index+1 exclusive.

I am unsure what you meant by this, so I did not make any changes here. The commit range is used for debugging and we consider the bounds of the range inclusive. Even when we store this information in the next PR we store it along side the last commit reference.

https://github.com/MystenLabs/sui/blob/5694132b59b366455eeb5db5967a703c55e2be9f/consensus/core/src/dag_state.rs#L596-L610

File: consensus/core/src/leader_scoring.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring.rs:234]: Mention that the round range is inclusive.

File: consensus/core/src/leader_scoring.rs
Comment (mwtian) [reply consensus/core/src/leader_scoring.rs]: Even if it is only used for debugging, it is strange that CommitRange is wrapping the Range type that has inclusive start and exclusive end, while we are using values with inclusive start and end. For example, if the first round among blocks is 5 and last round among blocks is 10, the correct CommitRange should be 5..11 or 5..=10, instead of 5..10.

File: consensus/core/src/leader_scoring.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring.rs:246]:
```suggestion
    pub(crate) fn find_supported_leader_block(
```

File: consensus/core/src/leader_scoring.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring.rs:263]: I would try to avoid recursion if possible. Otherwise the semantic when there are equivocations can be very subtle. Maybe add a TODO.

File: consensus/core/src/leader_scoring.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring.rs:303]: This should be fairly common because of weak links and committed ancestors.
```suggestion
                tracing::trace!(
```

File: consensus/core/src/leader_scoring_strategy.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring_strategy.rs:221]:
```suggestion
                tracing::trace!(
```

File: consensus/core/src/leader_scoring_strategy.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring_strategy.rs:225]:
```suggestion
                tracing::trace!(
```

File: consensus/core/src/leader_scoring_strategy.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring_strategy.rs:157]:
```suggestion
                    tracing::trace!(
```

File: consensus/core/src/leader_scoring_strategy.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring_strategy.rs:168]:
```suggestion
                tracing::trace!(
```

File: consensus/core/src/leader_scoring_strategy.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring_strategy.rs:171]:
```suggestion
                tracing::trace!(
```

File: consensus/core/src/leader_scoring_strategy.rs
Comment (mwtian) [inline consensus/core/src/leader_scoring_strategy.rs:95]: This will bias the scores to validators already with higher stake. Let's save this strategy else where and iterate on it more before checking it in.

File: consensus/core/src/leader_scoring.rs
Comment (arun-koshy) [reply consensus/core/src/leader_scoring.rs]: Aaah I see what you mean now, modified the related code to reflect this inclusive/exclusive range.

**Review by mwtian — APPROVED** [review]

**Pull request merged**
